# Notebook 01 - Data Dependencies

Data require for these analyses can be downloaded from [OSF.io](https://doi.org/10.17605/OSF.IO/YBS32).

**TODO**: add in links once data is uploaded

1. Read count data after `kneaddata` quality control ([link](#)) is used in Notebook 02.
1. Taxonomic profiles  ([link](#)) generated by `metaphlan` are used in Notebook 03.
1. Functional profiles generated by `humann` are used in Notebook 04.
   There are several functional tables that contain similar information,
   but in different forms. All are relative abundance normalized.
   1. `genefamilies_relab` ([link](https://osf.io/wfbxr/)):
      These files contain [UniRef90 ids](https://www.uniprot.org/help/uniref).
      These are the most complete, but the files are enormous.
   2. `pfam_names_relab` ([link](https://osf.io/eu48m/)):
      These files contain [protein family ids](http://pfam.xfam.org)
      and their human-readable names.
   3. `ko_names_relab` ([link](https://osf.io/736sv/)):
      These files contain [KEGG Orthology ids](https://www.genome.jp/kegg/ko.html)
      and their human-readable names.
   4. `ec_names_relab` ([link[](#)):](https://osf.io/jncwk/)
      These files contain [4-level enzyme commission numbers](https://enzyme.expasy.org)
      and their human-readable names.
2. Raw metagenomic sequences are not analyzed here directly,
   but can be acquired from the Sequence Read Archive (SRA) ([link](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA695570))

## Using this repository

Most of the code in this repository is written using the [julia programming language](http://julialang.org).
It has been tested on julia v1.6.

First, install julia by going to https://julialang.org/downloads/
and following the instructions for your operating system.

Next, download and unpack this repository from https://github.com/Klepac-Ceraj-Lab/ResonanceMicrobiome/releases,
or clone it using `git`.

```sh
curl -O https://github.com/Klepac-Ceraj-Lab/ResonanceMicrobiome/archive/refs/tags/v0.5.0.tar.gz
tar -xvzf v0.5.0.tar.gz
cd ResonanceMicrobiome
```

This repository contains files (`Project.toml` and `Manifest.toml`)
that enable you to exactly replicate the correct environment.
Once the directory is unpacked, make it your current working directory,
install the required packages by "instantiating" the project
from the julia REPL.

```
$ julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.6.0 (2021-03-24)
 _/ |\__'_|_|_|\__'_|  |
|__/                   |

julia> using Pkg; Pkg.activate()
  Activating environment at `~/repos/ResonanceMicrobiome/Project.toml`

julia> Pkg.instantiate()
```

Always activate this environment before running the code.
For scripts, this can be done by running with `julia --project=@. <script.jl>`,
or in the REPL by using `Pkg.activate` or the Pkg REPL mode.
[See here](https://docs.julialang.org/en/v1/manual/code-loading/#Project-environments)
for more information about using julia environments.

## Data Downloads

Data necessary to run this code will be automatically downloaded using DataDeps.jl.
The first time you try to use a dataset,
you will be asked if you want to download it.
Simply respond "yes", and the data will be downloaded and unpacked.
By default, data will be sent to a subdirectory in `~/.julia/datadeps`,
but this can be changed by setting the environmental variable `DATADEPS_LOAD_PATH`.

For example, to have data downloaded to your desktop, do something like:

```julia
julia> using ResonanceMicrobiome

julia> ENV["DATADEPS_LOAD_PATH"] = "/home/kevin/Desktop"
"/home/kevin/Desktop"

julia> datadep"taxonomic_profiles"
This program has requested access to the data dependency taxonomic_profiles.
which is not currently installed. It can be installed automatically, and you will not see this message again.

Taxonomic profiles generated by MetaPhlAn.



Do you want to download the dataset from https://osf.io/kyh48/download to "/home/kevin/Desktop/taxonomic_profiles"?
[y/n]
```

Note that you will need to set this environmental variable for each script
that uses this downloaded data,
but you can also set it in the shell by eg putting
`export DATADEPS_LOAD_PATH=/home/kevin/Desktop` in your `.envrc`.

